horovod/common/response_cache.cc

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "response_cache.h"
+
+namespace horovod {
+namespace common {
+
+void ResponseCache::clear() {
+  capacity_ = 0;
+  bits_outdated_ = false;
+  cache_.clear();
+  cache_iters_.clear();
+  tensor_name_to_bit_.clear();
+}
+
+void ResponseCache::set_capacity(uint32_t capacity) {
+  // Clear cache in case set_capacity is called multiple times if autotuning.
+  // Only clear if capacity is modified.
+  if (capacity != capacity_) {
+    this->clear();
+  }
+
+  capacity_ = capacity;
+  cache_iters_.reserve(capacity);
+}
+
+uint32_t ResponseCache::capacity() const { return capacity_; }
+
+size_t ResponseCache::current_size() const { return cache_.size(); }
+
+size_t ResponseCache::num_active_bits() const { return cache_iters_.size(); }
+
+ResponseCache::CacheState ResponseCache::cached(const Request& message) const {
+  auto it = tensor_name_to_bit_.find(message.tensor_name());
+  if (it != tensor_name_to_bit_.end()) {
+    // If entry associated with this request already exists in cache, check
+    // if tensor parameters match. If not, return that entry is invalid.
+    uint32_t cache_bit = it->second;
+    auto& cache_params = std::get<1>(*cache_iters_[cache_bit]);
+    return (cache_params.device == message.device() &&
+            cache_params.dtype == message.tensor_type() &&
+            cache_params.shape == message.tensor_shape())
+               ? CacheState::HIT
+               : CacheState::INVALID;
+  } else {
+    return CacheState::MISS;
+  }
+}
+
+ResponseCache::CacheState
+ResponseCache::cached(const Response& response,
+                      const TensorParams& params) const {
+  assert(response.tensor_names().size() == 1);
+  auto it = tensor_name_to_bit_.find(response.tensor_names()[0]);
+  if (it != tensor_name_to_bit_.end()) {
+    uint32_t cache_bit = it->second;
+    auto& cache_params = std::get<1>(*cache_iters_[cache_bit]);
+    return (cache_params.device == params.device &&
+            cache_params.dtype == params.dtype &&
+            cache_params.shape == params.shape)
+               ? CacheState::HIT
+               : CacheState::INVALID;
+  } else {
+    return CacheState::MISS;
+  }
+}
+
+void ResponseCache::put_(const Response& response, TensorParams& params) {
```

Comment:
Seems that there's a lot of logic spent maintaining consistency, but yet there's still `update_cache_bits()` required. 

What if we make this more batch-oriented, i.e. collect all responses this cycle, and once we call `update()`, new `cache_` will be created that is `new responses` + `responses from old cache_ that don't have same tensors as new responses`.  Then we don't need iterators and the process is super simple, right?